File: zul/src/org/zkoss/zul/api/Column.java
/* Column.java

{{IS_NOTE
	Purpose:
		
	Description:
		
	History:
		Tue Oct 22 09:27:29     2008, Created by Flyworld
}}IS_NOTE

Copyright (C) 2008 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under GPL Version 2.0 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
 */
package org.zkoss.zul.api;

import java.util.Comparator;
import org.zkoss.zk.ui.Components;//for javadoc
import org.zkoss.zk.ui.WrongValueException;

/**
 * A single column in a {@link Columns} element. Each child of the
 * {@link Column} element is placed in each successive cell of the grid. The
 * column with the most child elements determines the number of rows in each
 * column.
 * 
 * <p>
 * The use of column is mainly to define attributes for each cell in the grid.
 * 
 * <p>
 * Default {@link #getZclass}: z-column. (since 3.5.0)
 * 
 * @author tomyeh
 * @since 3.5.2
 */
public interface Column extends org.zkoss.zul.impl.api.HeaderElement {

	/** Returns the grid that contains this column. */
	public org.zkoss.zul.api.Grid getGridApi();

	/**
	 * Returns the sort direction.
	 * <p>
	 * Default: "natural".
	 */
	public String getSortDirection();

	/**
	 * Sets the sort direction. This does not sort the data, it only serves as
	 * an indicator as to how the grid is sorted.
	 * 
	 * <p>
	 * If you use {@link #sort(boolean)} to sort rows ({@link Row}), the sort
	 * direction is maintained automatically. If you want to sort it in
	 * customized way, you have to set the sort direction manaully.
	 * 
	 * @param sortDir
	 *            one of "ascending", "descending" and "natural"
	 */
	public void setSortDirection(String sortDir) throws WrongValueException;

	/** Sets the type of the sorter.
	 * You might specify either "auto", "auto(FIELD_NAME1[,FIELD_NAME2] ...)" or "none".
	 *
	 * <p>If "auto" is specified,
	 * {@link #setSortAscending} and/or {@link #setSortDescending} 
	 * are called with {@link org.zkoss.zul.RowComparator}, if
	 * {@link #getSortDescending} and/or {@link #getSortAscending} are null.
	 * If you assigned a comparator to them, it won't be affected.
	 * The auto created comparator is case-insensitive.
	 *
	 * <p>If "auto(FIELD_NAME1, FIELD_NAME2, ...)" is specified,
	 * {@link #setSortAscending} and/or {@link #setSortDescending} 
	 * are called with {@link org.zkoss.zul.FieldComparator}, if
	 * {@link #getSortDescending} and/or {@link #getSortAscending} are null.
	 * If you assigned a comparator to them, it won't be affected.
	 * The auto created comparator is case-insensitive.

	 * <p>If "none" is specified, both {@link #setSortAscending} and
	 * {@link #setSortDescending} are called with null.
	 * Therefore, no more sorting is available to users for this column.
	 * @since 3.5.3
	 */
	public void setSort(String type);

	/**
	 * Returns the ascending sorter, or null if not available.
	 */
	public Comparator getSortAscending();

	/**
	 * Sets the ascending sorter, or null for no sorter for the ascending order.
	 * 
	 * @param sorter
	 *            the comparator used to sort the ascending order. If you are
	 *            using the group feature, you can pass an instance of
	 *            {@link org.zkoss.zul.GroupComparator} to have a better
	 *            control. If an instance of
	 *            {@link org.zkoss.zul.GroupComparator} is passed,
	 *            {@link org.zkoss.zul.GroupComparator#compareGroup} is used to
	 *            group elements, and
	 *            {@link org.zkoss.zul.GroupComparator#compare} is used to sort
	 *            elements with a group. Otherwise, {@link Comparator#compare}
	 *            is used to group elements and sort elements within a group.
	 */
	public void setSortAscending(Comparator sorter);

	/**
	 * Sets the ascending sorter with the class name, or null for no sorter for
	 * the ascending order.
	 */
	public void setSortAscending(String clsnm) throws ClassNotFoundException,
			InstantiationException, IllegalAccessException;

	/**
	 * Returns the descending sorter, or null if not available.
	 */
	public Comparator getSortDescending();

	/**
	 * Sets the descending sorter, or null for no sorter for the descending
	 * order.
	 * 
	 * @param sorter
	 *            the comparator used to sort the ascending order. If you are
	 *            using the group feature, you can pass an instance of
	 *            {@link org.zkoss.zul.GroupComparator} to have a better
	 *            control. If an instance of
	 *            {@link org.zkoss.zul.GroupComparator} is passed,
	 *            {@link org.zkoss.zul.GroupComparator#compareGroup} is used to
	 *            group elements, and
	 *            {@link org.zkoss.zul.GroupComparator#compare} is used to sort
	 *            elements with a group. Otherwise, {@link Comparator#compare}
	 *            is used to group elements and sort elements within a group.
	 */
	public void setSortDescending(Comparator sorter);

	/**
	 * Sets the descending sorter with the class name, or null for no sorter for
	 * the descending order.
	 */
	public void setSortDescending(String clsnm) throws ClassNotFoundException,
			InstantiationException, IllegalAccessException;

	/**
	 * Sorts the rows ({@link Row}) based on {@link #getSortAscending} and
	 * {@link #getSortDescending}, if {@link #getSortDirection} doesn't matches
	 * the ascending argument.
	 * 
	 * <p>
	 * It checks {@link #getSortDirection} to see whether sorting is required,
	 * and update {@link #setSortDirection} after sorted. For example, if
	 * {@link #getSortDirection} returns "ascending" and the ascending argument
	 * is false, nothing happens. To enforce the sorting, you can invoke
	 * {@link #setSortDirection} with "natural" before invoking this method.
	 * Alternatively, you can invoke {@link #sort(boolean, boolean)} instead.
	 * 
	 * <p>
	 * It sorts the rows by use of {@link Components#sort}, if not live data
	 * (i.e., {@link Grid#getModel} is null).
	 * 
	 * <p>
	 * On the other hand, it invokes {@link org.zkoss.zul.ListModelExt#sort} to
	 * sort the rows, if live data (i.e., {@link Grid#getModel} is not null). In
	 * other words, if you use the live data, you have to implement
	 * {@link org.zkoss.zul.ListModelExt} to sort the live data explicitly.
	 * 
	 * @param ascending
	 *            whether to use {@link #getSortAscending}. If the corresponding
	 *            comparator is not set, it returns false and does nothing.
	 * @return whether the rows are sorted.
	 * @exception UiException
	 *                if {@link Grid#getModel} is not null but
	 *                {@link org.zkoss.zul.ListModelExt} is not implemented.
	 */
	public boolean sort(boolean ascending);

	/**
	 * Sorts the rows ({@link Row}) based on {@link #getSortAscending} and
	 * {@link #getSortDescending}.
	 * 
	 * @param ascending
	 *            whether to use {@link #getSortAscending}. If the corresponding
	 *            comparator is not set, it returns false and does nothing.
	 * @param force
	 *            whether to enforce the sorting no matter what the sort
	 *            direction ({@link #getSortDirection}) is. If false, this
	 *            method is the same as {@link #sort(boolean)}.
	 * @return whether the rows are sorted.
	 */
	public boolean sort(boolean ascending, boolean force);

	/**
	 * Groups and sorts the rows ({@link Row}) based on
	 * {@link #getSortAscending}. If the corresponding comparator is not set, it
	 * returns false and does nothing.
	 * 
	 * @param ascending
	 *            whether to use {@link #getSortAscending}. If the corresponding
	 *            comparator is not set, it returns false and does nothing.
	 * @return whether the rows are grouped.
	 */
	public boolean group(boolean ascending);

}

File: zul/src/main/resources/web/js/zul/grid/Column.ts
/* Column.ts

	Purpose:

	Description:

	History:
		Wed Dec 24 15:25:39     2008, Created by jumperchen

Copyright (C) 2008 Potix Corporation. All Rights Reserved.

This program is distributed under LGPL Version 2.1 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/
/**
 * A single column in a {@link Columns} element.
 * Each child of the {@link Column} element is placed in each successive
 * cell of the grid.
 * The column with the most child elements determines the number of rows
 * in each column.
 *
 * <p>The use of column is mainly to define attributes for each cell
 * in the grid.
 *
 * @defaultValue {@link getZclass}: z-column.
 */
@zk.WrapClass('zul.grid.Column')
export class Column extends zul.mesh.SortWidget {
	override parent!: zul.grid.Columns | undefined;

	/**
	 * @returns the grid that contains this column.
	 */
	getGrid(): zul.grid.Grid | undefined {
		return this.parent ? this.parent.parent : undefined;
	}

	constructor() {
		super(); // FIXME: params?
		this.listen({onGroup: this}, -1000);
	}

	/**
	 * @returns the rows of the grid that contains this column.
	 */
	getMeshBody(): zul.grid.Rows | undefined {
		var grid = this.getGrid();
		return grid ? grid.rows : undefined;
	}

	/** @internal */
	override checkClientSort_(ascending: boolean): boolean {
		const body = this.getMeshBody();
		return !(!body || body.hasGroup()) && super.checkClientSort_(ascending);
	}

	/**
	 * Groups and sorts the rows ({@link Row}) based on
	 * {@link getSortAscending}.
	 * If the corresponding comparator is not set, it returns false
	 * and does nothing.
	 *
	 * @param ascending - whether to use {@link getSortAscending}.
	 * If the corresponding comparator is not set, it returns false
	 * and does nothing.
	 * @param evt - the event causes the group
	 * @returns boolean whether the rows are grouped.
	 */
	group(ascending: boolean, evt: zk.Event): boolean {
		var dir = this.getSortDirection();
		if (ascending) {
			if ('ascending' == dir) return false;
		} else {
			if ('descending' == dir) return false;
		}

		var sorter = ascending ? this._sortAscending : this._sortDescending;
		if (sorter == 'fromServer')
			return false;
		else if (sorter == 'none') {
			evt.stop();
			return false;
		}

		var mesh = this.getMeshWidget();
		if (!mesh || mesh.isModel() || !zk.feature.pe || !zk.isLoaded('zkex.grid'))
			return false;
			// if in model, the sort should be done by server

		var	body = this.getMeshBody();
		if (!body) return false;
		evt.stop();

		var desktop = body.desktop,
			node = body.$n_();
		try {
			body.unbind();
			if (body.hasGroup()) {
				for (var gs = body.getGroups(), len = gs.length; --len >= 0;)
					body.removeChild(gs[len]);
			}

			interface Data {
				wgt: zk.Widget;
				index: number;
			}
			var d: Data[] = [],
				col = this.getChildIndex();
			for (var i = 0, z = 0, it = mesh.getBodyWidgetIterator(), w: zul.mesh.Item | undefined; (w = it.next()); z++)
				for (var k = 0, cell = w.firstChild; cell; cell = cell.nextSibling, k++)
					if (k == col) {
						d[i++] = {
							wgt: cell,
							index: z
						};
					}

			var dsc = dir == 'ascending' ? -1 : 1,
				fn = this.sorting,
				isNumber = sorter == 'client(number)';
			d.sort(function (a, b) {
				var v = fn(a.wgt, b.wgt, isNumber) * dsc;
				if (v == 0) {
					v = (a.index < b.index ? -1 : 1);
				}
				return v;
			});

			// clear all
			for (;body.firstChild;)
				body.removeChild(body.firstChild);

			for (var previous: Data | undefined, index = this.getChildIndex(), i = 0, k = d.length; i < k; i++) {
				const row = d[i];
				if (!previous || fn(previous.wgt, row.wgt, isNumber) != 0) {
					//new group
					let group: zkex.grid.Group,
						cell = row.wgt.parent!.getChildAt(index);
					if (cell && cell instanceof zul.wgt.Label) {
						group = new zkex.grid.Group();
						group.appendChild(new zul.wgt.Label({
							value: cell.getValue()
						}));
					} else {
						var cc = cell!.firstChild;
						if (cc && cc instanceof zul.wgt.Label) {
							group = new zkex.grid.Group();
							group.appendChild(new zul.wgt.Label({
								value: cc.getValue()
							}));
						} else {
							group = new zkex.grid.Group();
							group.appendChild(new zul.wgt.Label({
								value: msgzul.GRID_OTHER
							}));
						}
					}
					body.appendChild(group);
				}
				body.appendChild(row.wgt.parent!);
				previous = row;
			}
			this._fixDirection(ascending);
		} finally {
			body.replaceHTML(node, desktop);
		}
		return true;
	}

	override setLabel(label: string, opts?: Record<string, boolean>): this {
		super.setLabel(label, opts);
		if (this.parent)
			this.parent._syncColMenu();
		return this;
	}

	override setVisible(visible: boolean): this {
		if (this.isVisible() != visible) {
			super.setVisible(visible);
			if (this.parent)
				this.parent._syncColMenu();
		}
		return this;
	}

	/**
	 * It invokes {@link group} to group list items and maintain {@link getSortDirection}.
	 */
	onGroup(evt: zk.Event): void {
		var dir = this.getSortDirection();
		if ('ascending' == dir)
			this.group(false, evt);
		else if ('descending' == dir)
			this.group(true, evt);
		else if (!this.group(true, evt))
			this.group(false, evt);
	}

	/** @internal */
	override bind_(desktop?: zk.Desktop, skipper?: zk.Skipper, after?: CallableFunction[]): void {
		super.bind_(desktop, skipper, after);
		var n = this.$n_();
		this.domListen_(n, 'onMouseOver')
			.domListen_(n, 'onMouseOut');
		var btn = this.$n('btn');
		if (btn)
			this.domListen_(btn, 'onClick', '_doMenuClick');
	}

	/** @internal */
	override unbind_(skipper?: zk.Skipper, after?: CallableFunction[], keepRod?: boolean): void {
		var n = this.$n_();
		this.domUnlisten_(n, 'onMouseOver')
			.domUnlisten_(n, 'onMouseOut');
		var btn = this.$n('btn');
		if (btn)
			this.domUnlisten_(btn, 'onClick', '_doMenuClick');
		super.unbind_(skipper, after, keepRod);
	}

	/** @internal */
	_doMouseOver(evt: zk.Event): void {
		if (this.isSortable_()
				|| (this.parent!._menupopup && this.parent!._menupopup != 'none'))
			jq(this.$n_()).addClass(this.$s('hover'));
	}

	/** @internal */
	_doMouseOut(evt: zk.Event): void {
		if (this.isSortable_()
				|| (this.parent!._menupopup && this.parent!._menupopup != 'none')) {
			var $n = jq(this.$n_());
			if (!$n.hasClass(this.$s('visited')))
				$n.removeClass(this.$s('hover'));
		}
	}
}